libs/lib-postgres/src/locks/PostgresLockManager.ts

```diff
@@ -0,0 +1,126 @@
+import * as framework from '@powersync/lib-services-framework';
+import { v4 as uuidv4 } from 'uuid';
+import { DatabaseClient, sql } from '../db/db-index.js';
+
+const DEFAULT_LOCK_TIMEOUT = 60_000; // 1 minute
+
+export interface PostgresLockManagerParams extends framework.locks.LockManagerParams {
+  db: DatabaseClient;
+}
+
+export class PostgresLockManager extends framework.locks.AbstractLockManager {
+  constructor(protected params: PostgresLockManagerParams) {
+    super(params);
+  }
+
+  protected get db() {
+    return this.params.db;
+  }
+
+  get timeout() {
+    return this.params.timeout ?? DEFAULT_LOCK_TIMEOUT;
+  }
+
+  get name() {
+    return this.params.name;
+  }
+
+  async init() {
+    /**
+     * Locks are required for migrations, which means this table can't be
+     * created inside a migration. This ensures the locks table is present.
+     */
+    await this.db.query(sql`
+      CREATE TABLE IF NOT EXISTS locks (
+        name TEXT PRIMARY KEY,
+        lock_id UUID NOT NULL,
+        ts TIMESTAMPTZ NOT NULL
+      );
+    `);
+  }
+
+  protected async acquireHandle(): Promise<framework.LockHandle | null> {
+    const id = await this._acquireId();
+    if (!id) {
+      return null;
+    }
+    return {
+      refresh: () => this.refreshHandle(id),
+      release: () => this.releaseHandle(id)
+    };
+  }
+
+  protected async _acquireId(): Promise<string | null> {
+    const now = new Date();
+    const expiredTs = new Date(now.getTime() - this.timeout);
+    const lockId = uuidv4();
+
+    try {
+      // Attempt to acquire or refresh the lock
+      const res = await this.db.query(sql`
+        INSERT INTO
+          locks (name, lock_id, ts)
```

Comment:
[Postgres Advisory Locks](https://www.postgresql.org/docs/current/explicit-locking.html#ADVISORY-LOCKS) could be useful for this. It has the advantage that the lock will automatically be released when the connection is closed, e.g. due to the process crashing. That also means that you won't have any issues with refreshing the lock.

The disadvantage is that the lock is tied to a specific Postgres connection, so it may be more tricky to manage with connection pooling.

Either way, this is a bigger change, and we can investigate it after the initial release.

Reply:
Added a note to investigate after the initial release.